### PR TITLE
Add support for autowiring of services; apply to OAuth2 code

### PR DESCRIPTION
**Pull request by EreMaijala**
Adds support for autowiring services with dependencies that may be other services or VuFind configurations.

- Autowiring is opt-in; there must be an #[Autowire] attribute on the constructor.
- Configurations must be in array format and the attribute must indicate the configuration to inject.
- Service name is by default resolved from parameter type, but can be indicated with an additional `#[Autowire(service: 'name')] attribute.

Example constructor:

```
    /**
     * Constructor
     *
     * @param array       $mainConfig   Main VuFind configuration
     * @param array       $exportConfig Export-specific configuration
     * @param PhpRenderer $viewRenderer View renderer
     */
    #[Autowire]
    public function __construct(
        #[Autowire(config: 'config')]
        protected array $mainConfig,
        #[Autowire(config: 'export')]
        protected array $exportConfig,
        protected PhpRenderer $viewRenderer
    ) {
    }

```

TODO:
- [x] Decide if this is a useful thing to have.
- [x] Add support for autowiring different plugin classes (ILS drivers, database services etc.).
- [ ] Document autowiring in wiki.

### CODE REVIEW

**Comment by EreMaijala** [comment]
On my system with PHP 8.3.28 it takes about 2-4 microseconds to get the reflection data for the constructor. Trying to cache it would be slower. It seems that the only way this could be made faster would be an additional compilation step that would build service-specific factories ahead of time. The time to actually get the two configurations and the renderer in the example is typically around 350-600 microseconds (with extremes between 253 and 1665), so it seems that the reflection part adds very little.

It'd also be an option to use the reflection data for a utility that builds or updates an actual factory upon request. It would still require service manager configuration, though.

**Review by demiankatz — COMMENTED** [review]
Thanks, @EreMaijala, this is a great start! A few initial thoughts:

1.) I wonder if it would make sense to split this into two parts: an AbstractAutowiringFactory that checks for classes that are eligible for autowiring, and a non-abstract AutowiringFactory that builds an object using the annotations. This way, we could explicitly configure autowiring without having to rely on the magic of the abstract factory if we want to start using the functionality in the short term, and in the longer term, the generic abstract factory I revised in #4801 could just add autowiring support to its broader bag of tricks without having to run multiple abstract factories on every service creation.

2.) It might also make sense to soften the requirement that the class have the Autowire attribute -- we could use the attribute to detect classes that are eligible for autowiring, but we could still allow explicit autowiring of classes that don't contain the attribute.

3.) Regarding support for plugins, I think the easiest solution is to add a 'container' option to the autowire attribute -- then we can specify which plugin manager to pull the plugin from.

4.) It might be interesting to look at the Laminas [ConstructorParameterResolver](https://github.com/laminas/laminas-servicemanager/blob/4.6.x/src/Tool/ConstructorParameterResolver/ConstructorParameterResolver.php) for inspiration (this class isn't available until ServiceManager 4, but maybe we can use it directly in the future). At very least, I think its handling of default values for parameters might be worth incorporating into our version.

**Comment by EreMaijala** [comment]
@demiankatz Points 1 to 3 make sense to me. Regarding default values, I intentionally excluded default value handling. We have often used default values for new constructor params when we've wanted to maintain back-compatibility (with the expense of missing some functionality if they're not provided), and though they're probably now mostly properly typed, I didn't want us to automatically skip something that might still be untyped. At this point I believe it's better to require an explicit factory that could e.g. use the provided options to override default values. If we have cases where supporting the defaults would make sense, we could maybe require an attribute for it. This said, I'm ready to hear compelling arguments against these thoughts, of course. :)

**Comment by demiankatz** [comment]
> @demiankatz Points 1 to 3 make sense to me. Regarding default values, I intentionally excluded default value handling. We have often used default values for new constructor params when we've wanted to maintain back-compatibility (with the expense of missing some functionality if they're not provided), and though they're probably now mostly properly typed, I didn't want us to automatically skip something that might still be untyped. At this point I believe it's better to require an explicit factory that could e.g. use the provided options to override default values. If we have cases where supporting the defaults would make sense, we could maybe require an attribute for it. This said, I'm ready to hear compelling arguments against these thoughts, of course. :)

Yes, that makes sense -- we can leave default values alone until there's a use case! (Though maybe there's an argument for using default values if they are scalar types).

File: module/VuFind/src/VuFind/ServiceManager/Factory/Autowire.php [hunk 1]
```diff
@@ -47,17 +47,20 @@ class Autowire
...
     public function __construct(
         public readonly ?string $config = null,
         public readonly ?string $service = null,
+        public readonly ?string $container = null,
     ) {
         if (null !== $config && null !== $service) {
             throw new LogicException('#[Autowire] attribute cannot contain both config and service.');
         }
+        if (null !== $container && null === $service) {
+            throw new LogicException('#[Autowire] attribute cannot contain container without a service.');
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/ServiceManager/Factory/Autowire.php]: Is this restriction strictly necessary? If the argument is typed for a plugin found in the container, do we need to repeat that type as a service in the attribute, or could we just pull it from the type declaration? Maybe it's better to fail in the factory if we get a container and can't figure out a service name.

File: module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php [hunk 2]
```diff
@@ -115,24 +117,32 @@ protected function getConfigArray(ContainerInterface $container, string $config)
...
      * @SuppressWarnings(PHPMD.UnusedFormalParameter)
      */
-    protected function resolveServiceName(
+    protected function resolveService(
         ContainerInterface $container,
+        ?array $autowireArgs,
         ReflectionParameter $reflectionParameter
-    ): string {
+    ) {
+        if ($service = $autowireArgs['service'] ?? null) {
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php]: This doesn't seem right. Isn't `$autowireArgs['service']` the name of the service, not the service itself? I don't think returning it here makes sense -- shouldn't we be overriding the name instead?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala, see below for thoughts on the latest changes.

Maybe it's time to start building some tests to cover the edge cases.

File: module/VuFind/src/VuFind/ServiceManager/Factory/Autowire.php (hunk 1 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/ServiceManager/Factory/Autowire.php]: You're right, not necessary and removed.

File: module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php (hunk 2 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php]: I messed it up. Fixed, maybe.

**Comment by EreMaijala** [comment]
@demiankatz Yes, tests are next on the list.

File: module/VuFind/src/VuFind/ServiceManager/Factory/Autowire.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+     * @param string $container Container or plugin manager to use to get the service
+     */
+    public function __construct(
+        public readonly ?string $config = null,
+        public readonly ?string $service = null,
+        public readonly ?string $container = null,
+    ) {
+        if (null !== $config && null !== $service) {
+            throw new LogicException('#[Autowire] attribute cannot contain both config and service.');
+        }
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/ServiceManager/Factory/Autowire.php]: Should we also check for container here? Right now, container is meaningless if config is set.

```suggestion
        if (null !== $config && (null !== $service || null !== $container)) {
            throw new LogicException('#[Autowire] config attribute is incompatible with container or service attributes.');
        }
```

(I imagine we _could_ use the container attribute to override where the config is loaded from, but I'm not eager to overcomplicate this without a use case).

File: module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            $name = $type->getName();
+            if (null === $name || !($type instanceof ReflectionNamedType)) {
+                throw new LogicException('Unable to resolve type of parameter ' . $reflectionParameter->getName());
+            }
+        }
+        $containerToUse = ($containerName = $autowireArgs['container'] ?? null)
+            ? $container->get($containerName)
+            : $container;
+        return ($containerToUse instanceof PhpRenderer)
+            ? $containerToUse->plugin((string)$name)
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php]: I suppose this is harmless, but why would we use the renderer as a container when we could instead use the ViewHelperManager?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala! Looking much better; see below for a couple more minor things.

I've also been thinking about how to actually introduce this to the codebase. I wonder if one approach might be to start replacing the odd exceptional top-level service factories that couldn't be automatically replaced by #4801. If we can switch some of those to the autowiring factory instead, we can reduce the number of unconventional factories and also increase the benefit of the proposed abstract factory. (I haven't looked to see whether this is actually feasible in any/all cases, but maybe we could at least pick one or two of those as proof-of-concept cases).

File: module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php (hunk 4 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php]: It's just to mirror the many cases in current factories that get the ViewRenderer and request plugin from it. We can leave it out if you think it's better to not support that.

Comment (demiankatz) [reply module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php]: As long as we can retrieve view helpers directly from the helper plugin manager, I think it's better to do that than to unnecessarily construct a renderer just to pull plugins out of it. If there's some functionality that `plugin` has that is missing from the native plugin manager, then I could live with this -- but otherwise, I'd be inclined to say that if we're rewriting things to replace factories, we might as well just keep this in mind and simplify where necessary, rather than adding complexity to the factory.

File: module/VuFind/src/VuFindTest/ServiceManager/Factory/TestHarness/AutowiredClass.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+     *
+     * @param array      $config        Configuration
+     * @param Url        $url           URL helper
+     * @param Manager    $authManager   Authentication manager
+     * @param Connection $ilsConnection ILS Connection
+     */
+    public function __construct(
+        #[Autowire(config: 'config')]
+        protected array $config,
+        #[Autowire(container: 'ViewRenderer')]
```
Comment (demiankatz) [inline module/VuFind/src/VuFindTest/ServiceManager/Factory/TestHarness/AutowiredClass.php]: Should we switch this to ViewHelperManager, or do you want to wait and do some hands-on factory conversion first to see if there's a use case that really needs this?

**Review by demiankatz — COMMENTED** [review]
Tests look good! Just one question below related to our other open discussion thread.

File: module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php (hunk 4 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php]: Fair enough. I removed the special case completely. The HelperPluginManager is just another plugin manager with the normal get method, so doesn't need anything special. :)

File: module/VuFind/src/VuFindTest/ServiceManager/Factory/TestHarness/AutowiredClass.php (hunk 5 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFindTest/ServiceManager/Factory/TestHarness/AutowiredClass.php]: Special case dropped!

**Review by demiankatz — APPROVED** [review]
Thanks, @EreMaijala, this all looks good to me now. Do you want to merge it now so it's available, or should we figure out a way to actually apply it before we merge? (I lean toward finding an application before merging, but I can be persuaded to go the other route if you think it would be more helpful to take that approach).

**Comment by EreMaijala** [comment]
@demiankatz Well, that's a good question. I know I'd probably be happy to use when adding anything new, but I'm fine waiting until that moment is here.

**Comment by demiankatz** [comment]
@EreMaijala, I took a look at my idea of using autowiring to reduce remaining special cases in #4801. There are a number of factories that can't be easily replaced with autowiring, but the OAuth2 code looks like a promising place to start. See commit e913eaa for a quick proof-of-concept (I've confirmed that tests pass after the refactoring, and break if I use an inappropriate factory). If you like this, we could carry the work forward to deprecate/replace `VuFind\OAuth2\Repository\TokenRepositoryFactory` as well, and maybe that's enough of a start to justify merging this. What do you think?

**Comment by EreMaijala** [comment]
@demiankatz That looks good to me!

**Comment by EreMaijala** [comment]
@demiankatz Also if you encounter something that could be made easier with a smarter autowiring factory, please add a note.

**Review by demiankatz — APPROVED** [review]
Thanks, @EreMaijala! I have set up autowiring to replace the other "unconventional" OAuth2 factory in commit [ff25074](https://github.com/vufind-org/vufind/pull/4929/commits/ff25074cbd0f046a698f20ae4e5685e4df4a4655). I think this is probably a good enough excuse to merge the PR, but I'll leave it to you to push the button in case you have any last-minute concerns.

So far, this seems to cover all of the obvious use cases I can think of... but we'll see if we run into anything weirder!

**Comment by EreMaijala** [comment]
@demiankatz Looks good otherwise, but I started thinking that using YamlReader as container and config as service is really unintuitive. How about making it possible to use the `config` setting also for yaml? It could first try to get it from configManager, but if the result is empty, it could try to find a yaml instead. Then it would be ready for the day when ConfigManager also handles yaml.

**Comment by demiankatz** [comment]
> @demiankatz Looks good otherwise, but I started thinking that using YamlReader as container and config as service is really unintuitive. How about making it possible to use the `config` setting also for yaml? It could first try to get it from configManager, but if the result is empty, it could try to find a yaml instead. Then it would be ready for the day when ConfigManager also handles yaml.

I agree that might be more clear, but I hesitate to add a lot of fallback logic, both for performance reasons and because that can be unintuitive in a different way. How about this for a compromise:

Add a `configType` setting that can be `default` or `yaml`. When set to `yaml`, it appends `.yaml` on the config name and uses the `YamlReader` instead of the `ConfigManager`. This way, we can explicitly configure things, and in the future, if the `YamlReader` merges into the `ConfigManager`, no change will need to be made except ignoring the `configType` setting, which can be removed at that time.

**Comment by EreMaijala** [comment]
@demiankatz That's a good idea. I'll add it.

**Comment by EreMaijala** [comment]
@demiankatz Done, including a test case.

File: module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+     * @return array
+     */
+    protected function getConfigArray(
+        ContainerInterface $container,
+        ?array $autowireArgs,
+        string $config
+    ): array {
+        $type = $autowireArgs['configType'] ?? 'ini';
+        switch ($type) {
+            case 'ini':
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php]: One last thought -- I don't really like using `ini` as a value here, because if the config manager eventually becomes storage-agnostic, that value could become misleading/confusing.

I originally proposed `default` since that's very generic... but now I wonder if it would actually be better to have `array` and `object` values, so we have a wrapper around `$this->configManager->getConfigObject()`. I expect that `getConfigObject` will be deprecated soon, but it might be nice to have access to that method until that time, in case we want to start introducing more autowiring without having to drastically refactor legacy code.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks again, @EreMaijala -- I finally circled back around to this and had one last thought.

Also, do you plan to add the wiki page proposed in the TODO list, or do you want my help with that? I'm happy to help (with drafting, reviewing, proposing a location in the structure, or anything else you might need).

File: module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php (hunk 6 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/ServiceManager/Factory/AutowiringFactory.php]: That makes sense, now updated.

**Review by demiankatz — APPROVED** [review]
Thanks, @EreMaijala, I think this covers all the major use cases I can think of now!
